src/GraphQL/Definitions.hs

```diff
@@ -0,0 +1,222 @@
+  getFieldDefinitions = []
+
+-- symbols
+class GetSymbolList a where
+  getSymbolList :: [Prelude.String]
+
+instance forall a as. (KnownSymbol a, GetSymbolList as) => GetSymbolList (a:as) where
+  getSymbolList = (symbolVal (Proxy :: Proxy a)):(getSymbolList @as)
+
+instance GetSymbolList '[] where
+  getSymbolList = []
+
+
+-- object types from union type lists, e.g. for
+-- Union "Horse" '[Leg, Head, Tail]
+--               ^^^^^^^^^^^^^^^^^^ this part
+class UnionTypeObjectTypeDefinitionList a where
+  getUnionTypeObjectTypeDefinitions :: [ObjectTypeDefinition]
+
+instance forall a as. (HasObjectDefinition a, UnionTypeObjectTypeDefinitionList as) => UnionTypeObjectTypeDefinitionList (a:as) where
+  getUnionTypeObjectTypeDefinitions = (getDefinition @a):(getUnionTypeObjectTypeDefinitions @as)
+
+instance UnionTypeObjectTypeDefinitionList '[] where
+  getUnionTypeObjectTypeDefinitions = []
+
+
+-- Interfaces
+class HasInterfaceDefinitions a where
+  getInterfaceDefinitions :: Interfaces
+
+instance forall a as. (HasInterfaceDefinition a, HasInterfaceDefinitions as) => HasInterfaceDefinitions (a:as) where
+  getInterfaceDefinitions = (getInterfaceDefinition @a):(getInterfaceDefinitions @as)
+
+instance HasInterfaceDefinitions '[] where
+  getInterfaceDefinitions = []
+
+class HasInterfaceDefinition a where
+  getInterfaceDefinition :: InterfaceTypeDefinition
+
+instance forall ks fields. (KnownSymbol ks, HasFieldDefinitions fields) => HasInterfaceDefinition (Interface ks fields) where
+  getInterfaceDefinition =
+    let name = Name (toS (symbolVal (Proxy :: Proxy ks)))
+        in InterfaceTypeDefinition name (NonEmptyList (getFieldDefinitions @fields))
+
+-- Give users some help if they don't terminate Arguments with a Field:
+-- NB the "redundant constraints" warning is a GHC bug: https://ghc.haskell.org/trac/ghc/ticket/11099
+instance forall ks t. GHC.TypeLits.TypeError ('GHC.TypeLits.Text ":> Arguments must end with a Field") =>
+         HasFieldDefinition (Argument ks t) where
+  getFieldDefinition = undefined
+
+instance forall ks is ts. (KnownSymbol ks, HasInterfaceDefinitions is, HasFieldDefinitions ts) => HasAnnotatedType (Object ks is ts) where
+  getAnnotatedType =
+    let obj = getDefinition @(Object ks is ts)
+    in TypeNamed (DefinedType (TypeDefinitionObject obj))
+
+instance forall t ks. (KnownSymbol ks, HasAnnotatedType t) => HasFieldDefinition (Field ks t) where
+  getFieldDefinition =
+    let name = Name (toS (symbolVal (Proxy :: Proxy ks)))
+    in FieldDefinition name [] (getAnnotatedType @t)
+
+
+instance forall ks t b. (KnownSymbol ks, HasAnnotatedInputType t, HasFieldDefinition b) => HasFieldDefinition ((Argument ks t) :> b) where
+  getFieldDefinition =
+    let (FieldDefinition name argDefs at) = getFieldDefinition @b
+        argName = Name (toS (symbolVal (Proxy :: Proxy ks)))
+        arg = ArgumentDefinition argName (getAnnotatedInputType @t) Nothing
+    in (FieldDefinition name (arg:argDefs) at)
```

Comment:
I was the same, now I've changed my style slightly to use `let` when I'm just assigning names to sub-expressions that I am then using, and `where` mostly for multi-line, or generally more complex stuff. 

Can't quite describe the rule I have in my brain but it's clearly there because I can tell you when to use what :)